File: py2cpp/examples/csd.cpp
#include <iostream>
#include <string>
#include <cmath> // ceil, fabs, log
#include <string_view>

using std::ceil;
using std::fabs;
using std::log2;
using std::pow;
using std::string;

auto to_csd(double num, int places=0) -> string {
	if (num == 0.) return "0";
	auto absnum = fabs(num);
	auto n = absnum < 1. ? 0 : int(ceil(log2(absnum * 1.5)));
	auto csd_str = string{absnum < 1. ? "0" : ""};
	auto limit = pow(2., n) / 3.;
	while (n > -places) {
		if (n == 0) csd_str += '.';
		n -= 1;
		if (num > limit) {
		    csd_str += '+';
		    num -= 1.5 * limit;
		}
		else if (num < -limit) {
		    csd_str += '-';
		    num += 1.5 * limit;
		}
		else csd_str += '0';
		limit /= 2.;
	}
	return csd_str;
}


/** Convert the CSD string to a decimal */
auto to_decimal(std::string_view csd_str) -> double
{
    auto num = 0.0;
    auto loc = 0;
	auto i = 0;
    for (auto c : csd_str) {
		switch (c) {
        case '+':
            num *= 2.; num += 1.; break;
        case '-':
            num *= 2.; num -= 1.; break;
        case '0':
            num *= 2.; break;
        case '.':  // unlikely
            loc = i + 1;
		}
		++i;
	}            
    if (loc != 0)
        num /= std::pow(2, csd_str.size() - loc);
    return num;
}

double to_decimal_old(std::string csd_str){
	std::string left = "";
	std::string right = "";
	for(size_t i = 0;i<csd_str.size();i++){
		if(csd_str[i]=='.'){
			right = csd_str.substr(i+1);
			break;
		}
		else
			left += csd_str[i];
	}
   //std::cout << "left is:" << left << std::endl;
   //std::cout << "right is:" << right << std::endl;   
   int msb_power = left.size()-1;
   double num = 0.0;
   for(size_t i = 0;i<csd_str.size();i++){
	   double power_of_two = pow(2.0,(msb_power-(signed int)(i)));
	   if(csd_str[i] == '+')
		   num += power_of_two;
	   else if(csd_str[i] == '-')
		   num -= power_of_two;
	   //std::cout << "now msb_power-i is:" << msb_power-(signed int)(i) << std::endl;
   }
   std::cout << "num is :" << num << std::endl;
   return num;
}

auto to_csdfixed(double num, unsigned int nnz=4) -> string {
	if (num == 0.) return "0";
	auto an = fabs(num);
	auto n = an < 1. ? 0 : int(ceil(log2(an * 1.5)));
	auto csd_str = string{an < 1. ? "0" : ""};
    auto limit = pow(2., n) / 3.;
	while (n > 0 || nnz > 0) {
		if (n == 0) csd_str += '.';
		n -= 1;
		if (num > limit) {
			csd_str += '+';
			num -= 1.5 * limit;
			nnz -= 1;
		}
		else if (num < -limit) {
			csd_str += '-';
			num += 1.5 * limit;
			nnz -= 1;
		}
		else csd_str += '0';
		limit /= 2.;

		if (nnz == 0)
			num = 0;
	}
	return csd_str;
}
    

	
int main(void){
	std::string str1;
	double num1;
	str1 = to_csd(16.5);
	num1 = to_decimal("+0000.0+");
	return 0;
}

File: rust_by_examples/examples/csd.rs
/**
 Canonical Signed Digit Functions

 Handles:
  * Decimals
  *
  *

 eg, +00-00+000.0 or 0.+0000-00+
 Where: '+' is +1
        '-' is -1

 Harnesser
 License: GPL2
*/

/**
 * @brief Convert to CSD (Canonical Signed Digit) String representation
 *
 * Original author: Harnesser
 * https://sourceforge.net/projects/pycsd/
 * License: GPL2
 *
 * @param num
 * @param places
 * @return String
 */
fn to_csd(mut num: f64, places: i32) -> String {
    if num == 0.0 {
        return String::from("0");
    }
    let absnum = num.abs();
    let nn = (absnum * 1.5).log2().ceil() as i32;
    let mut n = if absnum < 1.0 { 0 } else { nn };
    let str = if absnum < 1.0 { "0" } else { "" };
    let mut csd_str = String::from(str);
    let mut pow2n = (2.0_f64).powf((n - 1) as f64);
    while n > -places {
        if n == 0 {
            csd_str.push('.');
        }
        n = n - 1;
        let d = 1.5 * num;
        if d > pow2n {
            csd_str.push('+');
            num = num - pow2n;
        } else if d < -pow2n {
            csd_str.push('-');
            num = num + pow2n;
        } else {
            csd_str.push('0');
        }
        pow2n = pow2n / 2.0;
    }
    csd_str
}

/**
 * @brief Convert the CSD String to a decimal
 *
 * @param csd_str
 * @return f64
 */
fn to_decimal(csd_str: &String) -> f64 {
    let mut num: f64 = 0.0;
    let mut loc: u32 = 0;
    let mut i: u32 = 0;
    for c in csd_str.chars() {
        if c == '0' {
            num = num * 2.0;
        } else if c == '+' {
            num = num * 2.0 + 1.0;
        } else if c == '-' {
            num = num * 2.0 - 1.0;
        } else if c == '.' {
            loc = i + 1;
        } // else unknown character
        i = i + 1;
    }
    if loc != 0 {
        num = num / (2.0_f64).powf(((csd_str.len() as u32) - loc) as f64);
    }
    return num;
}

/**
 * @brief Convert to CSD (Canonical Signed Digit) String representation
 *
 * @param[in] num
 * @param[in] nnz number of non-zero
 * @return String
 */
fn to_csdfixed(mut num: f64, mut nnz: u32) -> String {
    if num == 0.0 {
        return String::from("0");
    }
    let absnum = num.abs();
    let nn = (absnum * 1.5).log2().ceil() as i32;
    let mut n = if absnum < 1.0 { 0 } else { nn };
    let str = if absnum < 1.0 { "0" } else { "" };
    let mut csd_str = String::from(str);
    let mut pow2n = (2.0_f64).powf((n - 1) as f64);
    while n > 0 || (nnz > 0 && num.abs() > 1e-100) {
        if n == 0 {
            csd_str.push('.');
        }
        n = n - 1;
        let d = 1.5 * num;
        if d > pow2n {
            csd_str.push('+');
            num = num - pow2n;
            nnz = nnz - 1;
        } else if d < -pow2n {
            csd_str.push('-');
            num = num + pow2n;
            nnz = nnz - 1;
        } else {
            csd_str.push('0');
        }
        pow2n = pow2n / 2.0;
        if nnz == 0 {
            num = 0.0;
        }
    }
    csd_str
}

fn main() {
    println!("CHECK: +00-00.+0 == {}", to_csd(28.5, 2));
    println!("CHECK: 0.-0 == {}", to_csd(-0.5, 2));

    println!("CHECK: 28.5 == {}", to_decimal(&String::from("+00-00.+")));
    println!("CHECK: -0.5 == {}", to_decimal(&String::from("0.-")));

    println!("CHECK: +00-00.+ == {}", to_csdfixed(28.5, 4));
    println!("CHECK: 0.- == {}", to_csdfixed(-0.5, 4));
}
